[Auth] Update magic links for the new complete login update.

notandyvee opened the pull request:
### Fix

This PR makes some updates related to the magic link flow.

* Clicking login now doesn't show a bottom sheet, instead takes you directly to a login screen to enter your email.
* Here you have the option to enter your email and request a magic link. Or you can login with wordpress.
* If you click "Instantly Log In With Email", you will be taken to a screen to enter a code. The code is a backup to the magic link. In case the magic link doesn't work, the user can enter the code.

### Test

**Login with Wordpress**

- [ ] Click Log In.
- [ ] Click Log In with Wordpress.com
- [ ] This will pop up a web browser.
- [ ] Ensure at the end you are logged in.

The main reason to test this is because I update the deprecated `onActivityForResult`. It was just easier to handle it in the fragment.

**Login With Magic Link**

- [ ] Click Log In
- [ ] Enter your email and click "Instantly Log In With Email"
- [ ] Once you are on the "Enter Code" screen, go check your email.
- [ ] In your email, click the "Log In" button to deeplink into the app so you are auto logged in.
- [ ] You should be logged in.

**Login With Magic Link Code**

- [ ] Click Log In
- [ ] Enter your email and click "Instantly Log In With Email"
- [ ] Once you are on the "Enter Code" screen, go check your email.
- [ ] Copy the code from the email and go back to the app.
- [ ] Enter the code in the textfield and submit the code.
- [ ] This should login you in.


### Release

No Additional release update needed.

notandyvee committed:
[Auth] Update magic links for the new complete login update.

dangermattic commented:
<!--
  0 Errors
  2 Warnings: <code>strings.xml</code> files...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><code>strings.xml</code> files should only be updated on release branches, when the translations are downloaded by our automation.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

notandyvee committed:
Fix constraintlayout lint issue

khaykov commented on Simplenote/src/main/java/com/automattic/simplenote/authentication/SignInFragment.kt:
```diff
@@ -1,49 +1,137 @@
+
+    private var loginWithWordpress: Button? = null
+
+    private var mAuthState: String? = null
```
I would avoid mixing hungarian notation into kotlin :)

khaykov commented on Simplenote/src/main/java/com/automattic/simplenote/authentication/SignInFragment.kt:
```diff
@@ -1,49 +1,137 @@
+            val ctx = context ?: return@setOnClickListener
+            val authRequestBuilder = WordPressUtils.getWordPressAuthorizationRequestBuilder()
+
+            // Set unique state value.
```
Duplicate comment.

khaykov commented on Simplenote/src/main/java/com/automattic/simplenote/authentication/SignInFragment.kt:
```diff
@@ -1,49 +1,137 @@
 
-    private var loginWithPassword: TextView? = null
+
+    private var loginWithWordpress: Button? = null
```
Looks like we can remove loginWithWordpress from here, since it's only used to set a listener in inflateLayout.

khaykov commented on Simplenote/src/main/java/com/automattic/simplenote/authentication/SignInFragment.kt:
```diff
@@ -1,49 +1,137 @@
+
+        loginWithWordpress = view.findViewById(R.id.button_login_with_wordpress)
+        loginWithWordpress?.setOnClickListener {
+            val ctx = context ?: return@setOnClickListener
```
Can the context be null here? If so, should we use a context from the view instead?

khaykov commented on Simplenote/src/main/java/com/automattic/simplenote/authentication/magiclink/MagicLinkConfirmationFragment.kt:
```diff
@@ -0,0 +1,122 @@
+
+    private var progressDialogFragment: ProgressDialogFragment? = null
+
+    private var codeEditText: EditText? = null
```
Wonder if those View variables could be local?

notandyvee replied:
Yes to `codeEditText`. But no to `progressDialogFragment`. The reason being is we show and hide it from multiple places.

khaykov replied:
Yes, I only meant the views - codeEditText, actionCodeButton and loginWithPassword in this case:)

khaykov reviewed (commented):
I'm waiting for the rate limit to get better, but in meantime a couple of comments:

- When I enter the wrong code, the message says "Link no longer valid" which is misleading.

- After entering code I'm brought to the initial login screen:
![30e1c239e72723c346a3cb9141966aab](https://github.com/user-attachments/assets/16d975b4-c32c-4bb6-801b-5e97a0b27c72)

- Email delivery is pretty finicky, but is probably no the app's issue :)

- After rotating a screen when "Link no longer valid" dialog is visible, I get a "We're having problems" dialog appearing.
![b5e1f7674efeaee0a6c726c60ab2d290](https://github.com/user-attachments/assets/52e3b112-ee3c-496b-9f01-a253425a0683)

- There is no connection check when pressing Login button after entering code. With no internet I'm getting "we're having problems" toast.

khaykov commented:
Just a heads up - I tested with a regular, non cursed account, and both magic links and code work as expected. So I think we can ignore the "going to back to title screen after entering code" issue.

notandyvee committed:
PR review updates and fix some bugs.

roundhill reviewed (changes requested):
Looking good! I have a few UX requests to match up with iOS:

* Change `Instantly log in with email` button label to `Log in with email`

* Can you add the same `Or` and nice `Enter password` button on the code entry view?
![IMG_1930](https://github.com/user-attachments/assets/4a2ebb81-3ea7-42c7-9e60-d5c8b3503ebb)

Also one more nice thing iOS does is not require you to enter your email again if you select `Enter password`:

![IMG_5EF1CE57AC75-1](https://github.com/user-attachments/assets/c5bad9e1-c99c-4030-9a63-87744f3ee130)

notandyvee committed:
[Bug] Fix general state bugs with orientation changes.

notandyvee committed:
Requested PR review changes.

notandyvee committed:
Merge branch 'trunk' into andy/magic-link-v2

notandyvee commented:
@roundhill I did everything except:

> Also one more nice thing iOS does is not require you to enter your email again if you select Enter password:

There may be changes that need to be made with simperium. I'll confirm. So either way best to wait until we discuss this.

@khaykov ready for re-review. After your thorough review I found some small general fragment issues and viewmodel issues. I also updated some of the error messages. Again, keep in mind we have to balance security vs transparency here.

khaykov commented:
@notandyvee Thanks for the changes! I found one issue - when opening magic link, the app "closes" after logging you in:
![982fa3ba0a64a84049203b259cc63e24](https://github.com/user-attachments/assets/50543b37-8f13-4794-a9f0-c5cbe104da55)

notandyvee commented:
Thanks @khaykov . I'll investigate. The reason this is happening seems to be related to a bug I was fixing. Originally I would call `Simplenote#loginWithToken` and start `NotesActivity`. But doing so caused two `NoteActivity`s to pop up. During my testing it worked. But maybe I missed something.

notandyvee committed:
Removed need to pass Simplenote app class to magiclink viewmodel.

notandyvee committed:
Fix failing test.

notandyvee committed:
Re-added what I thought was not necessary.

notandyvee commented:
Hey @khaykov , so I think I fixed it. If you want to re-test. Idky since simperium is a bit of a black box for me. But I was able to reproduce what you encountered. The change I made should fix it. Let me know if you run into any problems.

roundhill reviewed (changes requested):
Working great, I have a few more UI requests if that's cool ;)

1. Can the line in the `Or` not extend out into the padding area? Try to make it flush with where the buttons end:
<img width="547" alt="Screenshot 2024-07-23 at 4 06 25 PM" src="https://github.com/user-attachments/assets/c080b799-a7ea-4ed1-ac43-c35a7c45ea12">

2. Update the top of the code view to have the `We sent a code to...` text like on iOS:

![IMG_1951](https://github.com/user-attachments/assets/56e27c5f-1df4-402a-8a55-4d3109c7fe9c)

3. Update the label `Login with password` here to say `Enter password`
<img width="318" alt="Screenshot 2024-07-23 at 4 10 01 PM" src="https://github.com/user-attachments/assets/9c80fc01-95dd-4e40-b9da-a8a4819e3014">

khaykov reviewed (approved):
Confirmed that the fix for opening link works 👍

notandyvee committed:
Tweak UI elements as they were off.

Also had to change how we set the title a bit in order to fix
a bug when switching between the sign in fragment and the
code confirmation fragment.

notandyvee committed:
Fix my goof.

roundhill reviewed (approved):
Working well, thanks for the UI updates :shipit:

notandyvee commented:
@roundhill changes requested made and and also fixed the issue with the code confirmation. Not sure what happened.

lol. sorry. Missed your message.
